doc/neps/nep-0056-array-api-main-namespace.rst

```diff
@@ -0,0 +1,524 @@
+change in ``rtol`` defaults for a linear algebra function), this will be
+handled the same way as any other backwards incompatible change in NumPy 2.0 -
+through documentation, release notes, and API migrations and deprecations over
+several releases.
+
+
+Detailed description
+--------------------
+
+In this section we'll focus on specific API additions and functionality that we
+would not consider introducing into NumPy if the standard would not exist and
+we wouldn't have to think/worry about its main goal of writing code that is
+portable across multiple array libraries and their supported features like GPUs
+and other hardware accelerators or JIT compilers.
+
+``device`` support
+^^^^^^^^^^^^^^^^^^
+
+Device support is perhaps the most obvious example. NumPy is and will remain a
+CPU-only library, so why bother introducing a ``ndarray.device`` attribute or
+``device=`` keywords in several functions? This one feature is purely meant to
+make it easier to write code that is portable across libraries. The ``.device``
+attribute will return an object representing CPU, and that object will be
+accepted as an input to ``device=`` keywords. For example:
+
+.. code::
+
+    # Should work when `xp` is `np` and `x1` a numpy array
+    x2 = xp.asarray([0, 1, 2, 3], dtype=xp.float64, device=x1.device)
+
+This will work as expected for NumPy, creating a 1-D numpy array from the input
+list. It will also work for CuPy & co, where it may create a new array on a GPU
+or other supported device.
+
+
+``isdtype``
+^^^^^^^^^^^
+
+The array API standard introduced a new function ``isdtype`` for introspection
+of dtypes, because there was no suitable alternative in NumPy. The closest one
+is ``np.issubdtype``, however that assumes a complex class hierarchy which
+other array libraries don't have, isn't the most ergonomic API, and required a
+larger API surface (``np.floating`` and friends). ``isdtype`` will be the new
+and canonical way to introspect dtypes. All it requires from a dtype is that
+``_eq__`` is implemented and has the expected behavior when compared with other
+dtypes from the same library.
+
+Note that as part of the effort on NEP 52, some dtype aliases were removed and
+canonical Python and C names documented. See also `gh-17325
+<https://github.com/numpy/numpy/issues/17325>`__ covering issues with NumPy's
+lack of a good API for this.
+
+
+``copy`` keyword semantics
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``copy`` keyword in ``asarray`` and ``array`` will now support
+``True``/``False``/``None`` with new meanings:
+
+- ``True`` - Always make a copy.
+- ``False`` - Never make a copy. If a copy is required a ``ValueError`` is raised.
+- ``None`` - A copy will only be made if it is necessary (previously ``False``).
+
+The ``copy`` keyword in ``astype`` will stick to its current meaning, because
+"never copy" when asking for a cast to a different dtype doesn't quite make
+sense.
```

Comment:
xref https://github.com/numpy/numpy/pull/25168#issuecomment-1857878408

It would be nice to get a little bit more discussion about how downstream should adapt to this change. Just replace `np.array(..., copy=False)` with `np.asarray()`? If so, do we communicate that in the `ValueError` that gets generated?

Reply:
Yes, this is a at least almost a large impact change:  It should have its own sub section in the backwards compatibility section.  (It is currently completely missing there).

Reply:
I assume you're commenting on `asarray`, not `astype`. I'd say the migration guide should say: do nothing (because that's more likely than not what the author intended - remember we have a semantic ambiguity here right now), and if you get an exception, inspect the code and if you indeed intended "only if needed", change it to `copy=None`.

I also don't think it's a large change. The default is do nothing, and in some cases the exception message will then tell you what to change.

Reply:
That is fair, although I still think it would be good to explicitly mention the BC change of having to replace `array(..., copy=False)` with `asarray(...)`.  You can point out that it isn't super common because `asarray()` makes more sense and that the it is turns into an easy to fix error.

I am actually not convinced that it's more likely than not what the author intended (unless you found that by code search).  Many cases are probably just better served by `asarray()` but tried to micro-optimize it away because it had extra overhead until a few years ago.
Also there is the caveat that if we go e.g. via `__array__()`, `copy=False` will fail even when it shouldn't conceptually.

Reply:
I've used `np.array(..., copy=False, subok=True)` quite a lot, when indeed it was faster than `np.asanyarray`; I don't think I have ever had a cases where I'd want an error if the input was not an array...

That said, I agree with the logic of `False/True/None` and don't mind to change my code/astropy (though I wish `copy=None` was backward compatible).

Reply:
> Also there is the caveat that if we go e.g. via `__array__()`, `copy=False` will fail even when it shouldn't conceptually.

This is more problematic, indeed.

Reply:
Quick comments while I'm sorting through the text and gh-25168:
1. there's a bit of a circular thing going on here, where most of the problem seems to be `array` (a non-recommended / don't care function), and the only reason we even want to update `array` is for consistency with `asarray` (the important function here)
2. I'm not 100% sure I understand the `__array__` comment. I haven't thought about that one much before, but I think you mean something here like *"we cannot know if a call to `__array__` is going to cause this non-numpy object to copy data, so we must not call `__array__` if `copy=False`"?* If so, I'm not sure that's necessarily a long-term problem; a logical solution would be to add `copy=None` keyword to `__array__` so that we can know. If numpy cannot know it preferably shouldn't be guessing.